src/Lights/Shadows/shadowGenerator.ts

```diff
@@ -1220,10 +1220,10 @@ export class ShadowGenerator implements IShadowGenerator {
 
             // Draw
             renderingMesh._processRendering(effectiveMesh, subMesh, effect, material.fillMode, batch, hardwareInstancedRendering,
-                (isInstance, worldOverride, effectiveMaterial, effectiveMeshOverride) => {
-                    if (effectiveMeshOverride && effectiveMesh !== effectiveMeshOverride) {
-                        effectiveMeshOverride.getMeshUniformBuffer().bindToEffect(effect, "Mesh");
-                        effectiveMeshOverride.transferToEffect(worldOverride);
+                (isInstance, worldOverride) => {
+                    if (effectiveMesh !== renderingMesh) {

```

Comment:
I think the condition should be `if (effectiveMesh !== renderingMesh && !isInstance) {` to be iso with the previous code.

Indeed, in the case where hardware instance rendering is not supported, `Mesh._processRendering` was calling the callback with `isInstance==true` and `effectiveMeshOverride==undefined` with the old code. In that case, we executed the `else` branch, so we should also do it with the new code.

Also, I think there's a bug in the old code, I think `effectiveMesh.transferToEffect(world);` should be `effectiveMesh.transferToEffect(isInstance ? worldOverride : world);` so that we use the world matrix passed to the function (which is the world matrix of the instance) when hardware instance rendering is not supported...

The relevant bit of code from `_processRendering` is:
```typescript
var world = instance.getWorldMatrix();
if (onBeforeDraw) {
    onBeforeDraw(true, world, effectiveMaterial);
}
```

Reply:
I'll trust your judgement on this. I have no idea what this code is supposed to do. I believe what I did is identical to what it did before. There were no checks for `isInstance` before.

Reply:
That said, I would prefer to do this change in a separate PR.

Reply:
The new check on `!isInstance` is to replace the old `effectiveMeshOverride` check. Previously we had:
```typescript
if (effectiveMeshOverride && effectiveMesh !== effectiveMeshOverride) {
```
`effectiveMesh !== effectiveMeshOverride` is now `effectiveMesh !== renderingMesh`  because `effectiveMeshOverride` is set to `renderingMesh` by the `Mesh._processRendering` calling code:
```typescript
if (onBeforeDraw) {
    onBeforeDraw(false, renderingMesh._effectiveMesh.getWorldMatrix(), effectiveMaterial, renderingMesh._effectiveMesh);
}
```
However, there's a second call to the callback in `Mesh._processRendering`:
```typescript
if (onBeforeDraw) {
    onBeforeDraw(true, world, effectiveMaterial);
}
```
With the old code in `ShadowGenerator`, we would take the `else` branch of the `if (effectiveMeshOverride && effectiveMesh !== effectiveMeshOverride) {` because `effectiveMeshOverride` is not set. In the new code, to differentiate the two `onBeforeDraw` calls, we can't use `effectiveMeshOverride` anymore but we can  use `isInstance` instead because the first call set it to `false` and the second to `true`. So, if `isInstance==true` we should take the `else` branch.

Reply:
I'll fix it in a separate PR.

Reply:
I think by disabling the hardware support for instances: `engine.getCaps().instancedArrays = false`.